src/viewport.js

```diff
@@ -514,34 +514,37 @@ $.Viewport.prototype = {
      * @param {OpenSeadragon.Rect} bounds
      * @returns {OpenSeadragon.Rect} constrained bounds.
      */
-     _applyBoundaryConstraints: function(bounds) {
-        var newBounds = new $.Rect(
-                bounds.x,
-                bounds.y,
-                bounds.width,
-                bounds.height);
+    _applyBoundaryConstraints: function(bounds) {
+        var newBounds = this.viewportToViewerElementRectangle(bounds).getBoundingBox();
```

Comment:
@iangilman Yes, exactly. The viewport rectangle needs to be rotated to match the viewer element, because the logic of the constraints is (and always was) implemented in terms of `x` and `y` directly. The rotated viewport bounds does not have the same `x` value for `topLeft` and `bottomLeft`, for example, because the corners are in the coordinate frame of the rotated image. If the logic was implemented as vector math relative to the line segments that define the edges of the rotated viewport rectangle, that would also work, but would be a huge refactor. In the previous implementation, which did not convert to viewer element coordinates, this was the reason for the unintuitive angled bouncing for non-90-degree rotations, and for the improper dimensions of the bounding area for 90-degree rotations.

Getting the bounding box for the rotated content is pretty obvious - we want to compute constraints from the min/max coordinates of the element versus the edges of the visible area. The reason for getting the bounding box of the viewport itself is less intuitive (isn't it already a rectangle?). The reason is that for some rotations, the coordinates of the rotated rectangle were automatically adjusting and making `topLeft` etc. incorrect, and the bounding box always is correct.

Reply:
Awesome... Makes good sense! Thank you for the thorough explanation. This is of course all part of the historic record now :)